src/transactions/cdp.js

```diff
@@ -16,748 +30,865 @@ const MINRATIO = 140;
-
-	let r1_stxns = [stxns[0].blob, stxn2.blob];
-	// txn3
-	let start = 0;
-	if ( !optedIn ) {
-		r1_stxns.push(stxns[2].blob)
-		start = 1;
-	}
-	if ((await getCurrentUnix()) - start_time > 30) {
-		return {
-			alert: true,
-			text: "Aborted: Transaction review and signing took too long. Please try again."
-		}
-	}
-
-	const sendTxn1Promise = sendTxn(r1_stxns)
-
-	lsig = algosdk.makeLogicSig(reserve.logic, [algosdk.encodeUint64(1)]);
-	let stxn4 = algosdk.signLogicSigTransactionObject(txn4, lsig);
-		
-	let stxns2 = [stxns[start + 2].blob, stxns[start + 3].blob, stxns[start + 4].blob, stxn4.blob]
-	let response = await sendTxn1Promise;
-	response = await sendTxn(stxns2, "Successfully opened a CDP with ID: " + accountID + ".")
-	updateCDP(info.address, accountID, openingMicroALGOs, microOpeningGard);
-	return response
-	// XXX: May want to do something else besides this, a promise? loading screen?
+export function createOptInTxn(params, info, assetID) {
+  params.fee = 1000;
+  console.log("creating opt in for ", assetID);
+  let txn = algosdk.makeAssetTransferTxnWithSuggestedParamsFromObject({
+    from: info.address,
+    to: info.address,
+    amount: 0,
+    suggestedParams: params,
+    assetIndex: assetID,
+  });
+  return txn;
 }
 
-export async function mint(accountID, newGARD) {
+export async function openCDP(openingALGOs, openingGARD) {
+  if (openingGARD < 1) {
+    alert(
+      "Opening GARD needs to be above 1.\n" +
+        "Your opening GARD is is: " +
+        openingGARD,
+    );
+    return null;
+  }
+
+  // Setting up promises
+  const infoPromise = accountInfo();
+  const microOpeningGard = microGARD(openingGARD);
+  const devFeesPromise = calcDevFeesCurrent(microOpeningGard, false);
+  const paramsPromise = getParams(2000);
+
+  // XXX: Could add a nice check here to make sure the ratio is acceptable
+
+  const openingMicroALGOs = parseInt(openingALGOs * 1000000);
+  const ratio = calcRatio(openingMicroALGOs, openingGARD);
+  if (ratio < MINRATIO) {
+    return {
+      alert: true,
+      text:
+        "Ratio needs to be above " +
+        MINRATIO +
+        "%.\n" +
+        "Your ratio is: " +
+        calcRatio(openingMicroALGOs, openingGARD, true),
+    };
+  }
+
+  // Part 1: Opting in, creating needed info, etc.
+
+  const info = await infoPromise;
+  const devFees = await devFeesPromise;
+  const accountIDPromise = findOpenID(info.address);
+
+  if (
+    307000 +
+      openingMicroALGOs +
+      100000 * (info["assets"].length + 1) +
+      devFees >
+    info["amount"]
+  ) {
+    return {
+      alert: true,
+      text:
+        "Depositing this much collateral will put you below your minimum balance.\n" +
+        "Your Maximum deposit is: " +
+        (info["amount"] -
+          devFees -
+          307000 -
+          100000 * (info["assets"].length + 1)) /
+          1000000 +
+        " Algos",
+    };
+  }
+  let optedIn = false;
+  let optedInGard = verifyOptIn(info, gardID);
```

Comment:
Could this be simplified by replacing all instances of "optedIn" with "optedInGard"? That's what optedIn referred to initially